File: bench/suite/src/st/mandelbrot/go/mandelbrot.go
// mandelbrot: single-threaded, same algorithm as the Benchmarks Game description.
// Writes a binary PBM (P4) bitmap of the Mandelbrot set to stdout.
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	n := 200
	if len(os.Args) > 1 {
		n, _ = strconv.Atoi(os.Args[1])
	}
	w, h := n, n
	const iter = 50
	const limit2 = 4.0
	out := bufio.NewWriterSize(os.Stdout, 1<<16)
	defer out.Flush()
	fmt.Fprintf(out, "P4\n%d %d\n", w, h)
	for y := 0; y < h; y++ {
		ci := 2.0*float64(y)/float64(h) - 1.0
		var b byte
		bitNum := 0
		for x := 0; x < w; x++ {
			cr := 2.0*float64(x)/float64(w) - 1.5
			var zr, zi, tr, ti float64
			for i := 0; i < iter && tr+ti <= limit2; i++ {
				zi = 2.0*zr*zi + ci
				zr = tr - ti + cr
				tr = zr * zr
				ti = zi * zi
			}
			b <<= 1
			if tr+ti <= limit2 {
				b |= 1
			}
			bitNum++
			if bitNum == 8 {
				out.WriteByte(b)
				b, bitNum = 0, 0
			} else if x == w-1 {
				b <<= uint(8 - w%8)
				out.WriteByte(b)
				b, bitNum = 0, 0
			}
		}
	}
}

File: bench/suite/src/st/mandelbrot/rust/mandelbrot.rs
// mandelbrot: single-threaded, same algorithm as the Benchmarks Game description.
// Writes a binary PBM (P4) bitmap of the Mandelbrot set to stdout.
use std::io::Write;

fn main() {
    let n: usize = std::env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(200);
    let (w, h) = (n, n);
    let iter = 50;
    let limit2 = 4.0f64;
    let row_bytes = (w + 7) / 8;
    let mut out = Vec::with_capacity(row_bytes * h + 32);
    write!(out, "P4\n{} {}\n", w, h).unwrap();
    for y in 0..h {
        let ci = 2.0 * y as f64 / h as f64 - 1.0;
        let mut byte: u8 = 0;
        let mut bit_num = 0;
        for x in 0..w {
            let cr = 2.0 * x as f64 / w as f64 - 1.5;
            let (mut zr, mut zi, mut tr, mut ti) = (0.0f64, 0.0f64, 0.0f64, 0.0f64);
            let mut i = 0;
            while i < iter && tr + ti <= limit2 {
                zi = 2.0 * zr * zi + ci;
                zr = tr - ti + cr;
                tr = zr * zr;
                ti = zi * zi;
                i += 1;
            }
            byte <<= 1;
            if tr + ti <= limit2 {
                byte |= 1;
            }
            bit_num += 1;
            if bit_num == 8 {
                out.push(byte);
                byte = 0;
                bit_num = 0;
            } else if x == w - 1 {
                byte <<= 8 - w % 8;
                out.push(byte);
                byte = 0;
                bit_num = 0;
            }
        }
    }
    std::io::stdout().write_all(&out).unwrap();
}
